server/leafnode.go

```diff
@@ -785,12 +790,17 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
 	return c
 }
 
-func (c *client) processLeafnodeInfo(info *Info) {
+func (c *client) processLeafnodeInfo(info *Info) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
 	if c.leaf == nil || c.isClosed() {
-		return
+		return nil
+	}
+
+	// Prevent connecting to client port.
+	if info.ClientConnectURLs != nil {
```

Comment:
It seems that a server will always have `info.LeafNodeURLs` will always have at least the URL of the server you connect to. So adding that check may help us detect both client and route ports.

Reply:
That being said, by default the route's port will send the INFO with `leafnode_urls` (except if leafnode disable advertise), so if client advertise is disabled but leafnode is not, then we would not know that we connect to the route. We could add a new field on the INFO that says which type of connection that is. But we would need to make sure that if the field is not set, not to consider it is wrong (would mean old server) and then that would be more reliable for an INFO that has this field set.

Reply:
That's right, though was not sure on whether to check if `leafnode_urls` were missing since seems there was a special condition that suggested that they could be missing here: https://github.com/nats-io/nats-server/blob/ecb5008fe33b5373e95183a6fc5fb68345a4a0eb/server/leafnode.go#L818
I suppose using `leafnode_urls` is better anyway because as a leafnode you would always advertise your own route? Maybe the both checks together would be a good heuristic?

Reply:
But that's why I am a bit uneasy with what we are checking in the INFO. I think that it may be safer to introduce a new field and obviously that will work only for new servers, but with current approach I would be afraid to reject a valid connection simply because there are conditions (that we missed) where those fields (leafnode_urls/client_urls) are not present.

Reply:
Ok, I think that a LEAF INFO will always have leafnode urls with at least the URL of that server. The ROUTE INFO may or may not have leafnode urls, depending of advertise and if that server itself has leafnode or not.
So I think it is safe to assume that an INFO from a LEAF port has always leafnode_urls, so in other words, the absence of it for sure means that it is not a LEAF port. But it is not enough to detect ROUTE port.

Reply:
Yes was not sure that checking for something missing in the INFO fields would work because all the conditions were not clear, so went for checking instead for `connect_urls` being present since that definitely means it is not meant for leafnodes, think that may cover many cases for now without changing the INFO protocol too much... agree maybe the right way would be to have an extra field to detect whether it is a leafnode connection, but also worried about backwards compatibility...